[정해성] sprint4

Sea0Soda opened the pull request:
##  기본 기능

- [x] 이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강 테두리와 아래에 “이메일을 입력해주세요.” 에러 메시지를 표시
- [x] 이메일 input에서 focus out 할 때, 이메일 형식이 잘못되었을 경우 “잘못된 이메일 형식입니다” 에러 메시지를 표시
- [x] 닉네임 input에서 focus out 할 때, 값이 없을 경우 “닉네임을 입력해주세요.” 에러 메시지를 표시
- [x] 비밀번호 input에서 focus out 할 때, 값이 없을 경우 “비밀번호를 입력해주세요.” 에러 메시지를 표시
- [x] 비밀번호 input에서 focus out 할 때, 8자 미만일 경우 “비밀번호를 8자 이상 입력해주세요.” 에러 메시지를 표시
- [x] 비밀번호 확인 input이 비밀번호와 다를 경우 “비밀번호가 일치하지 않습니다.” 에러 메시지를 표시
- [x] 하나라도 값이 비어있거나 에러가 있는 경우 ‘회원가입’ 버튼은 비활성화
- [x] 모든 입력이 유효한 경우 ‘회원가입’ 버튼이 활성화됨
- [x] 회원가입 버튼 클릭 시 로그인 페이지(`/login/index.html`)로 이동
- [x] 이메일 input focus out → 값 없음: “이메일을 입력해주세요.” 에러 표시
- [x] 이메일 형식이 잘못된 경우: “잘못된 이메일 형식입니다.” 에러 표시
- [x] 비밀번호 input focus out → 값 없음: “비밀번호를 입력해주세요.” 에러 표시
- [x] 비밀번호 8자 미만일 경우: “비밀번호를 8자 이상 입력해주세요.” 에러 표시
- [x] 하나라도 에러가 있으면 로그인 버튼 비활성화
- [x] 모든 입력이 유효하면 로그인 버튼 활성화
- [x] 로그인 버튼 클릭 시 `/items` 페이지로 이동

https://16sprintmission.netlify.app/
---

심화
- [ ] 눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 합니다.

- [ ] 비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이도록 합니다.


---

## 주요 변경사항

-
- 


---

## 스크린샷
![screencapture-127-0-0-1-5500-login-signup-html-2025-05-19-23_36_54](https://github.com/user-attachments/assets/b371e702-cd80-43c1-8954-c7df4bf33191)

![screencapture-127-0-0-1-5500-login-index-html-2025-05-19-23_36_31](https://github.com/user-attachments/assets/d4016d81-a88d-4b7a-a145-cb7079be3976)

---

## 멘토에게

- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.
-
-

Sea0Soda committed:
스프린트 미션4 1차 완성본

addiescode-sj commented on js/login.js:
```diff
@@ -0,0 +1,102 @@
+
+  emailInput.addEventListener('blur', () => {
+    const value = emailInput.value.trim()
+    const errorMsg = getOrCreateErrorElement(emailInput)
```
네이밍이 좀 더 명확해질 필요가 있겠네요. 함수 이름을 보고도 어떤 일을 하는지 예측하기 쉬운 네이밍을 사용해주세요 :)

함수 body를 보니까, 입력 필드에 대한 에러 메시지 요소를 찾고, 에러 UI를 추가하는 역할을 하는데 (getOrCreateErrorElement 네이밍처럼 찾거나 생성하는게 아니라, 찾으면 에러 UI 추가)

`validateErrorMessage` 정도는 어떨까요? :)

addiescode-sj commented on js/login.js:
```diff
@@ -0,0 +1,102 @@
+    const value = emailInput.value.trim()
+    const errorMsg = getOrCreateErrorElement(emailInput)
+
+    if (!value) {
+      showError(emailInput, errorMsg, '이메일을 입력해주세요.')
+    } else if (!isValidEmail(value)) {
+      showError(emailInput, errorMsg, '잘못된 이메일 형식입니다.')
+    } else {
+      clearError(emailInput, errorMsg)
+    }
```
하나의 이벤트 리스너안에서 너무 많은 일을 담당하고있어요.
유효성 검사의 경우 단일 책임을 가지고있는 함수로 분리해볼까요?

그리고 똑같은 if 조건문을 사용하더라도, early return을 잘 사용하면 복잡한 조건문 중첩을 줄일 수 있어 각 조건이 독립적으로 처리되고, 코드의 흐름 또한 명확해진답니다!

이런 방식으로 개선해보는건 어떨까요?

```js
const validateEmail = (value, input, errorElement) => {
  if (!value) {
    showError(input, errorElement, '이메일을 입력해주세요.');
    return;
  }
  
  if (!isValidEmail(value)) {
    showError(input, errorElement, '잘못된 이메일 형식입니다.');
    return;
  }
  
  clearError(input, errorElement);
};
```

addiescode-sj commented on js/login.js:
```diff
@@ -0,0 +1,102 @@
+
+  // 비밀번호 부분
+
+  passwordInput.addEventListener('blur', () => {
+    const value = passwordInput.value.trim()
+    const errorMsg = getOrCreateErrorElement(passwordInput)
+
+    if (!value) {
+      showError(passwordInput, errorMsg, '비밀번호를 입력해주세요.')
+    } else if (value.length < 8) {
+      showError(passwordInput, errorMsg, '비밀번호를 8자 이상 입력해주세요.')
+    } else {
+      clearError(passwordInput, errorMsg)
+    }
+    updateButtonState()
+  })
```
자세히 보면, emailInput의 이벤트리스너와 똑같은 구조로 코드가 짜여져있죠?
코드 중복을 줄이려면 어떻게 개선하는게 좋을까요? :)

addiescode-sj replied:
우선, 재사용 가능한 단위의 공통 로직을 단일 책임을 가진 함수로 추출해줍니다.
이 함수는 입력 필드에 대한 유효성 검사를 수행하는 함수로, 인자로 전달받은 입력필드와 검사를 수행할 타입을 체크해 유효성 검사를 수행합니다.

그러려면 프로그램 내 여기저기 흩어져있는 유효성 검사 규칙들을 한군데에서 관리하기 쉽게 모아주는게 좋겠죠?

아래 예시와 같이 유효성 검사 규칙은 객체로 만들어주고
해당 유효성 검사 규칙을 가지고 validateInput함수를 만들어봤습니다.

```js
const validationRules = {
  email: {
    validations: [
      {
        condition: (value) => !value,
        message: '이메일을 입력해주세요.'
      },
      {
        condition: (value) => !isValidEmail(value),
        message: '잘못된 이메일 형식입니다.'
      }
    ]
  },
  password: {
    validations: [
      {
        condition: (value) => !value,
        message: '비밀번호를 입력해주세요.'
      },
      {
        condition: (value) => value.length < 8,
        message: '비밀번호를 8자 이상 입력해주세요.'
      }
    ]
  }
};

const validateInput = (input, type) => {
  const value = input.value.trim();
  const errorMsg = getOrCreateErrorElement(input);
  const rules = validationRules[type];

  const failedValidation = rules.validations.find(
    ({ condition }) => condition(value)
  );

  if (failedValidation) {
    showError(input, errorMsg, failedValidation.message);
  } else {
    clearError(input, errorMsg);
  }
  
  updateButtonState();
};
```

추가적으로, 이 두가지는 프로그램을 이루는 핵심 규칙(비즈니스 로직)이다보니, 필요할때마다 재사용될수있겠죠?
export 키워드를 사용해 모듈화하시고 필요할때마다 import 받아서 사용하는 구조로 개선해보셔도좋습니다.

addiescode-sj replied:
그렇다면 이런 핵심 로직을 사용하는 페이지에서는
해당 함수를 이벤트 리스너를 등록해 올바로 사용만 해주면 되겠죠? :)

```js
// 이벤트 리스너 등록
emailInput.addEventListener('blur', () => validateInput(emailInput, 'email'));
passwordInput.addEventListener('blur', () => validateInput(passwordInput, 'password'));
```

addiescode-sj commented on js/login.js:
```diff
@@ -0,0 +1,102 @@
+      return error
+    }
+    return next
+  }
```
70, 71사이에 공백 한칸 띄워주세요! :)

addiescode-sj commented on js/signup.js:
여기서도 페이지마다 같은 코드를 복사 붙여넣기하듯 중복하는것보다는 모듈화를 시도해보시면 좋겠죠? :)

addiescode-sj reviewed (approved):
수고하셨습니다!
재사용 가능한 단위로 단일 책임을 가진 함수 만들고, 모듈화하는 팁 참고해보시고 리팩토링 진행해주시면 좋을것같아요.

## 주요 리뷰 포인트 
- 유지보수를 고려한 개발

Sea0Soda committed:
회원가입 및 로그인 유효성 검사 로직 모듈화 및 중복 제거

- validateInput 함수와 validationRules 객체를 활용해 공통 로직 분리
- 이메일, 비밀번호, 닉네임 등 조건 검사를 모듈에서 관리하도록 개선
- 중복된 blur 이벤트 로직 간소화
- 함수 이름(getOrCreateErrorElement → validateErrorMessage) 통일로 가독성 향상
